### PR TITLE
Fix missing alpha in Windows DIB images

### PR DESCRIPTION
**Pull request by prehisle**
## Summary

Some Windows screenshot tools put 32-bit DIB clipboard data on the clipboard with RGB pixels populated but alpha set to 0 for every pixel. When RustDesk forwards that image to Linux as `image/png`, consumers such as Pinta interpret the paste as fully transparent and show a blank image.

This treats all-zero alpha with non-empty RGB content as missing alpha for Windows DIB clipboard data and restores alpha to 255. Images with any real alpha data are left unchanged.

## Context

Observed with WeChat Alt+A screenshots sent through RustDesk to Ubuntu/Pinta: the received PNG had valid RGB data, but alpha extrema were `(0, 0)`. PixPin screenshots did not trigger the issue. This is consistent with RustDesk issue rustdesk/rustdesk#10062 and mg-chao/snow-shot#817.

## Testing

- Added unit coverage for the alpha repair helper.
- `git diff --check` passes locally.
- `cargo fmt` / `cargo test` were not run locally because this environment does not have a Rust toolchain installed; please rely on CI/Windows test run for compilation.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved clipboard image handling: after converting to RGBA the app now classifies header alpha behavior and repairs missing/ambiguous alpha. Declared transparency is preserved; truly missing alpha is made fully opaque; an "unknown" case uses a RGB-present/alpha-zero heuristic to restore visible pixels.

* **Tests**
  * Added unit and integration tests verifying opaque restoration, preservation of declared transparency, and the RGB-present/alpha-zero heuristic.

### CODE REVIEW

**Comment by rustdesk** [comment]
@fufesou

**Comment by fufesou** [comment]
Thanks for the fix. Confirmed that the transparent image issue is solved.

I think the `CF_DIBV5` fallback still needs another pass, though.

It only fixes the "non‑zero RGB + zero alpha" case. A completely black screenshot (RGB also zero) with missing alpha would still decode as fully transparent. That feels like the same bug.

There's also a small compatibility risk because `CF_DIBV5` carries explicit bitmap metadata in `BITMAPV5HEADER` (`bV5Compression`, `bV5AlphaMask`). I think we should check that header metadata (or the source format if possible) before falling back to pixel heuristics. If we decide alpha is missing just by looking at the pixel buffer, a legitimate all‑transparent `DIBV5` image with non‑zero RGB channels would become opaque. That may be less common for screenshot tools, but `arboard` is still decoding a general Windows clipboard format here.

Could we instead look at the `DIBV5` header/format metadata where possible, and keep the pixel scan only as an explicit fallback? At the very least, I think we should add tests for both edge cases:

- all‑alpha‑zero, all‑black RGB – should that become opaque for the screenshot‑tool case?

- all‑alpha‑zero, non‑zero RGB, but intended as a transparent image – should that be preserved?

- A `read_cf_dibv5()`‑level test, not just a helper‑level test.



https://learn.microsoft.com/en-us/windows/win32/api/wingdi/ns-wingdi-bitmapv5header

https://learn.microsoft.com/en-us/windows/win32/dataxchg/standard-clipboard-formats

**Comment by prehisle** [comment]
Thanks, that makes sense. I updated the PR to make the DIBV5 path header-first:

- If bV5AlphaMask is non-zero, alpha is treated as explicitly present and preserved.
- If the header uses BI_RGB or BI_BITFIELDS with no alpha mask, alpha is treated as missing and normalized to 255 for every pixel. This covers the all-black screenshot case too.
- The RGB/alpha pixel scan is now only used for the Unknown metadata case.

I also added read_cf_dibv5()-level tests for:

- all-black BI_RGB pixels with missing alpha becoming opaque
- BI_BITFIELDS pixels with an alpha mask preserving fully transparent RGB data

Validation note: this machine still has no Rust toolchain, so I could only run git diff --check locally.

File: src/platform/windows.rs [hunk 1]
```diff
@@ -470,6 +541,125 @@ mod image_data {
...
+		data.extend_from_slice(pixels);
+		data
+	}
+
+	#[test]
+	fn read_cf_dibv5_repairs_all_black_missing_alpha() {
+		let data = test_dibv5(2, 1, BI_RGB, 0, &[0, 0, 0, 0, 0, 0, 0, 0]);
+		let image = read_cf_dibv5(&data).unwrap();
+
+		assert_eq!(image.width, 2);
```
Comment (fufesou) [inline src/platform/windows.rs]: Compilation fails here. Maybe you should
```rust
     let ImageData::Rgba(image) = read_cf_dibv5(&data).unwrap() else {
         panic!("expected RGBA image");
     };
```

File: src/platform/windows.rs [hunk 2]
```diff
@@ -470,6 +541,125 @@ mod image_data {
...
+
+	#[test]
+	fn pixel_heuristic_preserves_fully_empty_pixels() {
+		let mut data = [0; 12];
+
+		assert!(!repair_missing_alpha_if_opaque_rgb(&mut data));
+		assert_eq!(data, [0; 12]);
+	}
+
+	fn test_dibv5(
```
Comment (fufesou) [inline src/platform/windows.rs]: This is a pure test helper. You can add `#[test]` above.

**Comment by prehisle** [comment]
Good catch. This fork uses ImageData as an enum, so the read_cf_dibv5() tests need to destructure ImageData::Rgba before checking width/height/bytes. I fixed that in 1b0415d.

I also installed a local Rust toolchain and ran:

- rustfmt --check src/platform/windows.rs
- cargo check
- cargo check --tests --target x86_64-pc-windows-gnu
- git diff --check

Those pass.

I also ran cargo test on Linux. It compiles, but two existing clipboard/environment-dependent tests fail here: tests::get_set_special due env_logger already being initialized, and tests::all_tests because the current clipboard data differs from the expected PNG bytes. These failures do not involve the Windows DIBV5 code path.

**Comment by fufesou** [comment]
The helper-level tests feel a bit too detailed now that the real `read_cf_dibv5()` path is covered. I would keep the tests that prove the external behavior:
   - missing alpha is repaired for a DIBV5 without an alpha mask, including the all-black case
   - declared alpha is preserved when the DIBV5 header has an alpha mask

The lower-level helper tests can probably be reduced to one small fallback/unknown-header case, or removed if the `read_cf_dibv5()` tests already cover the intended behavior.

Otherwise, the tests are coupled to the current helper split rather than the observable clipboard decoding behavior.

File: src/platform/windows.rs (hunk 2 above)
Comment (prehisle) [reply src/platform/windows.rs]: Thanks. Since this helper takes parameters and returns a Vec<u8>, adding #[test] directly would make it an invalid test function. I marked it #[cfg(test)] instead and renamed it from test_dibv5 to dibv5_test_data so it is test-only without looking like a standalone test.

**Comment by prehisle** [comment]
Agreed. I trimmed the helper-level tests in 6e4ff6c so the PR now focuses on the observable read_cf_dibv5() behavior:

- DIBV5 without an alpha mask, including all-black pixels, is decoded as opaque
- DIBV5 with a declared alpha mask preserves transparent pixels

I kept only one small helper-level test for the Unknown metadata fallback path, since that branch is intentionally not covered by the normal BI_RGB/BI_BITFIELDS read_cf_dibv5() cases.

Re-ran:

- rustfmt --check src/platform/windows.rs
- cargo check --target x86_64-pc-windows-gnu
- cargo check --tests --target x86_64-pc-windows-gnu
- git diff --check

**Comment by fufesou** [comment]
Perhaps it would be better to introduce a test mod instead of adding `#[cfg(test)]` in different places.

like what RustDesk does

https://github.com/rustdesk/rustdesk/blob/cc7fe4efdca7b2ac38a0f40f45e852c04eb70edc/src/platform/windows.rs#L4507

**Comment by prehisle** [comment]
Good point. I moved the Windows image tests and their fixture helper into a nested #[cfg(test)] mod tests in dce2fac, so the production module no longer has scattered test-only annotations.

Re-ran:

- rustfmt --check src/platform/windows.rs
- cargo check --target x86_64-pc-windows-gnu
- cargo check --tests --target x86_64-pc-windows-gnu
- git diff --check

**Review by fufesou — APPROVED** [review]

File: src/platform/windows.rs [hunk 3]
```diff
@@ -449,26 +519,112 @@ mod image_data {
 		}
 	}
 
-	#[test]
-	fn conversion_between_win_and_rgba() {
-		const DATA: [u8; 16] =
-			[100, 100, 255, 100, 0, 0, 0, 255, 255, 100, 100, 255, 100, 255, 100, 100];
+	#[cfg(test)]
+	mod tests {
```
Comment (fufesou) [inline src/platform/windows.rs]: Maybe you can move this mod to the end of the file.

Comment (prehisle) [reply src/platform/windows.rs]: Done. I moved the test module to the end of the file and kept the DIBV5 helper visibility private by limiting the file-level tests to the observable read_cf_dibv5() behavior. Re-ran rustfmt --check src/platform/windows.rs, cargo check --target x86_64-pc-windows-gnu, cargo check --tests --target x86_64-pc-windows-gnu, and git diff --check.

File: src/platform/windows.rs [hunk 4]
```diff
@@ -448,28 +518,6 @@ mod image_data {
 			ImageDataCow::Owned(u32pixels_buffer)
 		}
 	}
-
-	#[test]
-	fn conversion_between_win_and_rgba() {
```
Comment (fufesou) [inline src/platform/windows.rs:453]: Sorry, there's still one last change.
Please move this test into `mod test {...}`.

File: src/platform/windows.rs [hunk 5]
```diff
@@ -449,26 +519,31 @@ mod image_data {
...
-		const DATA: [u8; 16] =
-			[100, 100, 255, 100, 0, 0, 0, 255, 255, 100, 100, 255, 100, 255, 100, 100];
+	#[cfg(test)]
+	mod tests {
+		use super::*;
+
+		#[test]
+		fn conversion_between_win_and_rgba() {
```
Comment (fufesou) [inline src/platform/windows.rs]: I mean, you can move this function into `mod tests` at the end of the file, not wrapping it in `mod tests` here.

Comment (prehisle) [reply src/platform/windows.rs]: Done. I moved conversion_between_win_and_rgba() into the file-level #[cfg(test)] mod tests at the end of the file. Only rgba_to_win() and win_to_rgba() are now pub(super) so that file-level test can call them; the DIBV5 alpha helpers remain private.

Re-ran rustfmt --check src/platform/windows.rs, cargo check --target x86_64-pc-windows-gnu, cargo check --tests --target x86_64-pc-windows-gnu, and git diff --check.

**Review by fufesou — APPROVED** [review]
